### PR TITLE
Invoice repository

### PR DESCRIPTION
**Pull request by netia1128**
# Description

This pull request is to incorporate the addition of Invoice and InvoiceRepository classes as well as their respective spec tests. 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Been Tested?

Tested and found to pass all spec tests and spec harness tests.

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [X] !!!!!Every method has a test!!!!
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules
- [X] Require 'simplecov' 
- [X] SimpleCov.start
- [X] Used require_relative 

#Test coverage

32.6%

### CODE REVIEW

File: lib/invoice.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  attr_reader   :id,
+                :customer_id,
+                :merchant_id,
+                :created_at,
+                :invoice_repo
+
+  attr_accessor :status,
+                :updated_at
+
+  def initialize(row, invoice_repo)
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply lib/invoice.rb:14]: It seems like this is saying that there are too many lines of code in the initialize. Based on the specs, we cannot avoid this.

File: lib/invoice.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+require 'bigdecimal'
+require 'time'
+
+class Invoice
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply lib/invoice.rb:4]: Discuss with teachers to see if this is necessary

File: lib/invoice.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+require 'bigdecimal'
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply lib/invoice.rb:1]: Discuss with teachers to see what this is.

File: lib/invoice_repository.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    new_id = highest_id_invoice.id + 1
+  end
+
+  def create(attributes)
+    attributes[:id] = generate_new_id
+    @invoices << Invoice.new(attributes, self)
+  end
+
+  def update(id, attributes)
+    if find_by_id(id) != nil && attributes[:status] != nil
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply lib/invoice_repository.rb:64]: This is fixed.

File: lib/invoice_repository.rb [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    @invoices.find_all do |invoice|
+      invoice.status == status.to_sym
+    end
+  end
+
+  def generate_new_id
+    highest_id_invoice = @invoices.max_by do |invoice|
+      invoice.id
+    end
+    new_id = highest_id_invoice.id + 1
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply lib/invoice_repository.rb:55]: This is fixed.

File: lib/invoice_repository.rb [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  end  
+  
+  def find_all_by_status(status)
+    @invoices.find_all do |invoice|
+      invoice.status == status.to_sym
+    end
+  end
+
+  def generate_new_id
+    highest_id_invoice = @invoices.max_by do |invoice|
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply lib/invoice_repository.rb:52]: Not sure how to implement this, will discuss with team.

File: lib/invoice_repository.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+      invoice.customer_id == customer_id
+    end
+  end
+
+  def find_all_by_merchant_id(merchant_id)
+    @invoices.find_all do |invoice|
+      invoice.merchant_id == merchant_id
+    end
+  end  
+
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply lib/invoice_repository.rb:44]: Fixed

File: lib/invoice_repository.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
+require_relative 'sales_engine'
+require_relative 'invoice'
+
+class InvoiceRepository
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply lib/invoice_repository.rb:4]: Speak with instructor.

File: lib/invoice_repository.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
+require_relative 'sales_engine'
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply lib/invoice_repository.rb:1]: Speak with instructor.

File: lib/sales_engine.rb [hunk 10]
```diff
@@ -23,8 +24,7 @@ def initialize(item_csv_location, merchant_csv_location)
   def self.from_csv(csv_hash)
     item_csv_location = csv_hash[:items]
     merchant_csv_location = csv_hash[:merchants]
-    SalesEngine.new(item_csv_location, merchant_csv_location)
+    invoice_csv_location = csv_hash[:invoices]
+    SalesEngine.new(item_csv_location, merchant_csv_location, invoice_csv_location)
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply lib/sales_engine.rb:29]: Ignore per instruction.

File: spec/invoice_repository_spec.rb [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        id: 5000,
+        customer_id: 2,
+        merchant_id: 3,
+        created_at: Time.now
+        }
+      test_invoice = ir.find_by_id(1)
+      ir.update(1, attributes)
+      expect(test_invoice.id).to eq(1)
+      expect(test_invoice.customer_id).to eq(1)
+      expect(test_invoice.merchant_id).to eq(12335938)
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply spec/invoice_repository_spec.rb:154]: Ignore per group discussion.

File: spec/invoice_repository_spec.rb [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+        id: '1234567890',
+        customer_id: '456789',
+        merchant_id: '234567890',
+        status: 'pending',
+        created_at: '2016-01-11 11:51:37 UTC',
+        updated_at: '1993-09-29 11:56:40 UTC'
+      }
+      ir.create(attributes)
+      expected = ir.find_by_id(4986)
+      expect(expected.merchant_id).to eq(234567890)
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply spec/invoice_repository_spec.rb:136]: Ignore per group discussion.

File: spec/invoice_repository_spec.rb [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    it 'updates invoices attributes' do
+      mock_sales_engine = instance_double('SalesEngine')
+      ir = InvoiceRepository.new('./data/invoices.csv', mock_sales_engine)
+      attributes = {
+        status: :success,
+        id: 5000,
+        customer_id: 2,
+        merchant_id: 3,
+        created_at: Time.now
+        }
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply spec/invoice_repository_spec.rb:149]: Fixed.

File: spec/invoice_repository_spec.rb [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+        status: 'pending',
+        created_at: '2016-01-11 11:51:37 UTC',
+        updated_at: '1993-09-29 11:56:40 UTC'
+    },
+    ir
+    )
+      ir.invoices << test_invoice1
+      ir.invoices << test_invoice2
+      expect(ir.find_all_by_merchant_id(234567890)).to eq([test_invoice1, test_invoice2])
+      expect(ir.find_all_by_merchant_id(123456789099999999)).to eq([])
```
[the remark this answers is not in the document]
Comment (ztrokey) [reply spec/invoice_repository_spec.rb:111]: Ignore per group discussion.

File: lib/sales_engine.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (netia1128) [reply lib/sales_engine.rb:29]: Put line break after each comma

File: lib/invoice_repository.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (netia1128) [reply lib/invoice_repository.rb:52]: Ignore this

**Pull request merged**